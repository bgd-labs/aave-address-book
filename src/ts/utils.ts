import {Address, getAddress} from 'viem';
import * as addresses from './AaveAddressBook';

type AnyObject = {[key: string]: any};

function flattenObject(obj: AnyObject, parentKey = '', result: AnyObject = {}): AnyObject {
  for (const [key, value] of Object.entries(obj)) {
    const newKey = parentKey ? `${parentKey}.${key}` : key;

    if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
      flattenObject(value, newKey, result);
    } else {
      result[newKey] = value;
    }
  }
  return result;
}

/**
 * Checks if address is listed on address-book
 * @param value
 * @param chainId
 * @returns string[] found paths to address-book addresses
 */
export function getAddressBookReferences(value: Address, chainId: number): string[] {
  // glob imports have no object properties
  // therefore we recreate the object via spread & remove addresses unrelated to the chain we are checking
  const transformedAddresses = Object.keys(addresses).reduce(
    (acc, key) => {
      if (addresses[key as keyof typeof addresses].CHAIN_ID === chainId) {
        const chainAddresses = {...addresses[key as keyof typeof addresses]};
        // deleting eModes as they only contain duplicates and on lookup should always be deprioritized
        if ((chainAddresses as any).E_MODES) {
          delete (chainAddresses as any).E_MODES;
        }
        return {...acc, ...flattenObject({[key]: chainAddresses})};
      }
      return acc;
    },
    {} as Record<string, Address>,
  );
  return Object.keys(transformedAddresses).filter(
    (key) => transformedAddresses[key] === getAddress(value),
  );
}
