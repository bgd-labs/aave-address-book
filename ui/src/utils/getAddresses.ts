import * as addressBook from '../../../src/ts/AaveAddressBook';
import { isAddress, Address, zeroAddress } from 'viem';

export type ListItem = {
  path: string[];
  value: string;
  chainId: number;
};

export function flattenObject(
  obj: any,
  path: string[] = [],
  chainId: number | null = null,
): ListItem[] {
  const result: ListItem[] = [];
  const entries = Object.entries(obj).sort(([keyA], [keyB]) => {
    if (keyA === 'CHAIN_ID') return -1;
    if (keyB === 'CHAIN_ID') return 1;
    return 0;
  });

  for (let [key, value] of entries) {
    if (key === 'E_MODES') continue;

    const newPath = [...path, key];
    if (key === 'CHAIN_ID') {
      chainId = value as number;
    }
    if (typeof value === 'object' && value !== null) {
      result.push(...flattenObject(value, newPath, chainId));
    } else if (isAddress(value as string) && value !== zeroAddress) {
      result.push({
        path: newPath,
        value: value as Address,
        chainId: chainId!,
      });
    }
  }
  return result;
}

export const flattenedAddresses = flattenObject(addressBook);
