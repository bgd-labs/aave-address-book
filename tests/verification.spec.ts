import {ChainId, ChainList} from '@bgd-labs/rpc-env';
import {describe, expect, it} from 'vitest';
import {flattenedAddresses, ListItem} from '../ui/src/utils/getAddresses';
import verified from './cache/verified.json';
import {writeFileSync} from 'fs';
import {zeroAddress} from 'viem';

const ETHERSCAN_API_KEY = process.env.ETHERSCAN_API_KEY as string;

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function checkProxyVerification(item: ListItem, guid: string) {
  const params = {
    chainId: String(item.chainId),
    apikey: ETHERSCAN_API_KEY,
    module: 'contract',
    action: 'checkproxyverification',
    guid,
  };
  const formattedParams = new URLSearchParams(params).toString();
  const url = `${getApiUrl(item.chainId)}?${formattedParams}`;
  try {
    const request = await fetch(url);
    const {status} = await request.json();
    if (status === '1')
      console.log(
        'successfully verified proxy. Please rerun the script in a few minutes as you need to wait till etherscan prunes its cache.',
      );
  } catch (e) {}
}

async function verifyProxy(item: ListItem) {
  const params = {
    chainId: String(item.chainId),
    apikey: ETHERSCAN_API_KEY,
    module: 'contract',
    action: 'verifyproxycontract',
  };
  const formattedParams = new URLSearchParams(params).toString();
  const url = `${getApiUrl(item.chainId)}?${formattedParams}`;
  try {
    const request = await fetch(url, {
      method: 'POST',
      body: new URLSearchParams({
        address: item.value,
      }),
    });

    const {status, result} = await request.json();
    await sleep(1000);
    if (status === '1') checkProxyVerification(item, result);
  } catch (e) {
    console.error(e);
    process.exit(1);
  }
}

async function checkVerified(item: ListItem) {
  const params = {
    chainId: String(item.chainId),
    apikey: ETHERSCAN_API_KEY,
    address: item.value,
    module: 'contract',
    action: 'getsourcecode',
  };

  const formattedParams = new URLSearchParams(params).toString();
  const url = `${getApiUrl(item.chainId)}?${formattedParams}`;
  try {
    const request = await fetch(url);
    const {status, result} = await request.json();
    if (status !== '1' || !result[0].ContractName) {
      // etherscan returns proxy contracts as non verified if the proxy is not manually assigned
      // therefore we try to manually assign it
      if (['S_TOKEN', 'V_TOKEN', 'A_TOKEN'].includes(item.path[item.path.length - 1])) {
        await verifyProxy(item);
      }
    }
    return {status, result};
  } catch (e) {
    console.error(e);
    return {status: '0', result: e};
  }
}

function getApiUrl(chainId: number) {
  if (chainId === ChainId.metis)
    return `https://api.routescan.io/v2/network/mainnet/evm/1088/etherscan/api`;
  return `https://api.etherscan.io/v2/api`;
}

describe(
  'verification',
  () => {
    it('should have all contracts verified except for the known set of errors', async () => {
      const addressesToCheck = flattenedAddresses.filter(
        (item) =>
          ![ChainId.harmony, ChainId.fantom].includes(item.chainId as any) &&
          !ChainList[item.chainId].testnet,
      );
      const errors: {item: ListItem}[] = [];
      let newVerified = false;
      // unique set of addresses checked on this iteration
      // used to prevent double checking the same address
      const checked = new Set<string>();
      for (const item of addressesToCheck) {
        const hasBeenCheckedBefore = verified[item.chainId][item.value];
        if (!hasBeenCheckedBefore && item.value !== zeroAddress) {
          const key = `${item.chainId}-${item.value}`;
          if (checked.has(key)) continue;
          checked.add(key);
          const {status, result} = (await checkVerified(item)) as {
            status: string;
            result: {ContractName: string}[];
          };
          await sleep(300);
          if (status !== '1' || !result[0].ContractName) {
            errors.push({item});
            console.log(result);
          } else {
            newVerified = true;
            if (!verified[item.chainId]) verified[item.chainId] = {};
            verified[item.chainId][item.value] = {
              name: result[0].ContractName,
            };
          }
        }
      }
      if (newVerified) {
        writeFileSync('./tests/cache/verified.json', JSON.stringify(verified, null, 2), {
          encoding: 'utf-8',
        });
      }
      expect(errors).toMatchSnapshot();
    });
  },
  {timeout: 120_000},
);
