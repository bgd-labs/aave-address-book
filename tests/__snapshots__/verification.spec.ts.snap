// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`verification > should have all contracts verified except for the known set of errors 1`] = `
[
  {
    "item": {
      "chainId": 1101,
      "path": [
        "GovernanceV3PolygonZkEvm",
        "PC_DATA_HELPER",
      ],
      "value": "0xF1c11BE0b4466728DDb7991A0Ac5265646ec9672",
    },
  },
  {
    "item": {
      "chainId": 1,
      "path": [
        "AaveV2EthereumAMM",
        "UI_INCENTIVE_DATA_PROVIDER",
      ],
      "value": "0xD01ab9a6577E1D84F142e44D49380e23A340387d",
    },
  },
  {
    "item": {
      "chainId": 137,
      "path": [
        "AaveV2Polygon",
        "UI_INCENTIVE_DATA_PROVIDER",
      ],
      "value": "0x645654D59A5226CBab969b1f5431aA47CBf64ab8",
    },
  },
  {
    "item": {
      "chainId": 43114,
      "path": [
        "AaveV2Avalanche",
        "UI_INCENTIVE_DATA_PROVIDER",
      ],
      "value": "0x11979886A6dBAE27D7a72c49fCF3F23240D647bF",
    },
  },
]
`;
